### ISSUE TITLE
Type hinting `NumericArray`, `Matrix3x3Like` and `Matrix4x4Like`

### ISSUE THREAD

**Issue by tkoyama010** [pyvista/pyvista#5074]
### Describe what maintenance you would like added.

1. Is the type hinting for `NumericArray` correct? Currently, as it's defined doing `Sequence[NumericArray]` implies that mixing and matching `Tuples` with `NDArray` is allowed, something like `((1,2,3), np.array([4,5,6]))`. I don't think that's generally expected or handled though... Perhaps it should be something like this instead: 
```python 
NumericArray = Union[Sequence[Number], Sequence[Sequence[Number]], np.ndarray]
```
2. Also, perhaps there should be a `Matrix3x3Like` type and `Matrix4x4Like` type as well, something like:
``` python
NumericArray = Union[Sequence[Number], Sequence[Sequence[Number]], np.ndarray]
Matrix3x3Like = Union[VectorArray, vtkMatrix3x3] 
Matrix4x4Like = Union[NumericArray, vtkMatrix4x4, vtkTransform] 
TransformLike = Union[Matrix3x3Like, Matrix4x4Like]
```

_Originally posted by @user27182 in https://github.com/pyvista/pyvista/pull/5052#discussion_r1355901047_

**Comment by darikg** [comment]
> Currently, as it's defined doing Sequence[NumericArray] implies that mixing and matching Tuples with NDArray is allowed, something like ((1,2,3), np.array([4,5,6])). I don't think that's generally expected or handled though... 

Seems like that would be fine if it's just going to be passed into `np.asarray` -- `np.asarray(((1,2,3), np.array([4,5,6])))` correctly returns a 2x3 array

**Comment by user27182** [comment]
I'd like to clarify what the issue is. Currently, the following type definition:
https://github.com/pyvista/pyvista/blob/3f8b06f0ee7b05dacf177a10f6219b77e57ac8ff/pyvista/core/_typing_core.py#L11
is missing the `Sequence[Sequence[Number]]` case, i.e. it fails to capture 2D arrays created using nested sequences. I would consider  ``[[1,2],[3,4]]`` to be a numeric array, but it's not a `NumericArray` based on the current definition.

One possible solution proposed by @tkoyama010 [here](https://github.com/pyvista/pyvista/pull/5052#discussion_r1354066563) is to use `Sequence[NumericArray]` to capture this missing case. This is what my comment (reposted above) is responding to.

> Seems like that would be fine if it's just going to be passed into `np.asarray` -- `np.asarray(((1,2,3), np.array([4,5,6])))` correctly returns a 2x3 array

Fair enough! If it quacks then it must be a duck. I still think it's a rare use case though, and removing it from the typing won't pose a risk of causing any type hint errors. 

I don't really think it's good practice to be mixing and matching these types together... either use sequences (1D) (then cast as numpy array), nested sequences (2D) (then cast as numpy array), or use a numpy array directly.  Since `np.array` is happy to cast just about anything as an array, I would personally _prefer_ that a type hint error be issued if I tried to use `Sequence[Sequence | np.ndarray]`

From a readability, English-language perspective I also think the naming is odd. The term `Sequence[NumericArray]` does not make sense to me intuitively, and in fact suggests that I _should_ be using a `Sequence` instead of using a `np.ndarray`directly to represent a 2D array. Therefore, to get the correct typing we would actually need to define yet _another_ type to capture this case, e.g. 
``` python
# current definition
NumericArray = Union[Sequence[Number], np.ndarray]

# captures nested Sequence, mix/match Sequence with ndarray, and general ndarray
NumericArray2D = Union[Sequence[NumericArray], np.ndarray]
```

I think all of this is confusing just to capture an edge case that I don't think should really be used even if it's possible. So, my original suggestion stands, which is to define it as:
```python
NumericArray = Union[Sequence[Number], Sequence[Sequence[Number]], np.ndarray]
```

**Comment by MatthewFlamm** [comment]
I _think_ that `NumericArray` is supposed to represent a 1D array.  It could be a case of bad naming of the variable.

**Comment by MatthewFlamm** [comment]
The problem we have is that we generally accept array_like, using numpy terminology, inputs from users.  To type this correctly we need to use typing that typically includes `Sequence[float]` for example and `ndarray`.  The purpose of the `_coerce_points_like_args` function, for example, was to coerce an array_like to ndarray along with some checking to ensure it has the right shape. In newer numpy versions we could use `ArrayLike` and `NDArray`, but it doesn't solve all problems as `ArrayLike` doesn't distinguish based on the type of data inside.  We'd also have to pin a minimum numpy version.

**Comment by user27182** [comment]
> I _think_ that `NumericArray` is supposed to represent a 1D array. It could be a case of bad naming of the variable.

If that's the case then yes that would an appropriate definition for a 1D array. Doing a search of the code reveals the type is barely used at all. Here's one case:
https://github.com/pyvista/pyvista/blob/3f8b06f0ee7b05dacf177a10f6219b77e57ac8ff/pyvista/core/utilities/arrays.py#L59

but for points, why not do `Union[Vector, VectorArray]` instead? Since we already have:
https://github.com/pyvista/pyvista/blob/3f8b06f0ee7b05dacf177a10f6219b77e57ac8ff/pyvista/core/_typing_core.py#L8

**Comment by user27182** [comment]
As a general comment, I am working on some typing improvements for PyVista that will be bundled together with #4977, along with a bunch of new validation methods. I think we should be defining all commonly used numeric inputs of various shapes explicitly, and use `typing.Annotated` to provide medatadata about the expected shape which can be used at runtime to validate input.

This would look something like

``` python
# FILE: _typing_core.py
from numpy.typing import NDArray
from collections import namedtuple 
from typing import Annotated, Sequence, Tuple, TypeVar, Union

# Define shape metadata
_MetaData = namedtuple('_MetaData', ['shape'])
_shape2 = _MetaData((2,))
_shape3 = _MetaData((3,))
_shapeN = _MetaData((-1,))
_shape3x3 = _MetaData((3, 3))
_shape4x4 = _MetaData((4, 4))
_shapeNx3 = _MetaData((-1, 3))
_shapeNxN = _MetaData((-1, -1))

# Define types
Number = Union[int, float]
_Numeric = Union[np.integer, np.floating]
NumericDType = TypeVar("NumericDType", bound=_Numeric)

_Sequence2 = Annotated[Sequence[Number], _shape2]
_NDArray2 = Annotated[NDArray[NumericDType], _shape2]
ArrayLike2 = Union[_Sequence2, _NDArray2]

_Sequence3 = Annotated[Sequence[Number], _shape3]
_NDArray3 = Annotated[NDArray[NumericDType], _shape3]
ArrayLike3 = Union[_Sequence3, _NDArray3]

_SequenceN = Annotated[Sequence[Number], _shapeN]
_NDArrayN = Annotated[NDArray[NumericDType], _shapeN]
ArrayLikeN = Union[_Sequence3, _NDArray3]

_Sequence3x3 = Annotated[Sequence[Sequence[Number]], _shape3x3]
_NDArray3x3 = Annotated[NDArray[NumericDType], _shape3x3]
ArrayLike3x3 = Union[_Sequence3x3, _NDArray3x3]

_Sequence4x4 = Annotated[Sequence[Sequence[Number]], _shape4x4]
_NDArray4x4 = Annotated[NDArray[NumericDType], _shape4x4]
ArrayLike4x4 = Union[_Sequence4x4, _NDArray4x4]

_SequenceNx3 = Annotated[Sequence[Sequence[Number]], _shapeNx3]
_NDArrayNx3 = Annotated[NDArray[NumericDType], _shapeNx3]
ArrayLikeNx3 = Union[_SequenceNx3, _NDArrayNx3]

_SequenceNxN = Annotated[Sequence[Sequence[Number]], _shapeNxN]
_NDArrayNxN = Annotated[NDArray[NumericDType], _shapeNxN]
ArrayLikeNxN = Union[_SequenceNxN, _NDArrayNxN]
```

A public API for the types would then look something like:
```python
# FILE: typing.py

from ._typing_core import (
    Number,
    ArrayLike2,
    ArrayLike3,
    ArrayLike3x3,
    ArrayLike4x4,
    ArrayLikeNx3,
    ArrayLikeNxN,
    BoundsLike,
    Matrix3x3Like,
    Matrix4x4Like,
    TransformLike,
)

```

> In newer numpy versions we could use `ArrayLike` and `NDArray`, but it doesn't solve all problems as `ArrayLike` doesn't distinguish based on the type of data inside. We'd also have to pin a minimum numpy version.

If support for older versions of numpy is needed, we could stuff the expected numeric type as part of the metadata for the annotation and forgo the use of NDArray and ArrayLike.

**Comment by MatthewFlamm** [comment]
I think the complexity of the points like data typing is because of a lack of cohesion in the codebase. Different functions that take a points like data as an argument do things differently..  Some functions only accept points that are 2D, i.e. (n, 3) shaped.  Some functions accept points that are either 2D or 1D, i.e. (n, 3) or (3,).  In the case of data with shape (3,) it is converted to (1, 3).

**Comment by user27182** [comment]
> I think the complexity of the points like data typing is because of a lack of cohesion in the codebase. Different functions that take a points like data as an argument do things differently.. Some functions only accept points that are 2D, i.e. (n, 3) shaped. Some functions accept points that are either 2D or 1D, i.e. (n, 3) or (3,). In the case of data with shape (3,) it is converted to (1, 3).

I agree with this. As a solution I think types should be very generic, steer away from terms like `Vector` and `Points`, and just specify the allowed shape, e.g. for points do `Union[ArrayLike3, ArrayLikeNx3]` as the allowed input type, then use a generic function like `coerce_as_ArrayLikeNx3` (i.e. effectively the same as  _coerce_pointslike_arg but with generic name).

**Comment by MatthewFlamm** [comment]
This typing problem for numeric data is likely not to be solved here.  I would probably err on the side of choosing the fewest number of types possible to represent the data.  For example, just 1 type variable that represents all array like data of any shape.  This is something like 
`Union[Sequence[Union[float, int]], Sequence[Sequence[Union[float, int]]], ndarray]`

**Comment by user27182** [comment]
> This typing problem for numeric data is likely not to be solved here. 

I agree the actual type hinting, in terms of an IDE or mypy generating an error, won't be solved here. Still, using `Annotated` can be useful for input validation as the numeric type or shape can be passed with this kind of typing. Perhaps I introduced confusion here by mixing "typing for input validation" with general correct typing hints in PyVista.

As a side note, the many types I gave earlier can be made fully private and only used internally for input validation purposes, if desired, or could even be removed completely, since the types would just be used as a short-hand to do `check_input_matches_type(user_input, expected_type)` where `expected_type ` could be `Union[ArrayLike3, ArrayLikeNx3])`. But, I realize this could also be done more generically as well with something like `check_input_matches_type(user_input, expected_shape, expected_dtype)`, in which case none of the specific ArrayLike types I suggested are necessary.

> I would probably err on the side of choosing the fewest number of types possible to represent the data. For example, just 1 type variable that represents all array like data of any shape. This is something like
> `Union[Sequence[Union[float, int]], Sequence[Sequence[Union[float, int]]], ndarray]`

This is a good point and would make things clearer by getting rid of any "middle" types.

I may have steered the discussion off-topic with the input validation... I mostly wanted to highlight the link between the current issue and the need to also validate the types being hinted.
